src/basic/pidfd-util.c

```diff
@@ -145,3 +147,26 @@ int pidfd_verify_pid(int pidfd, pid_t pid) {
 
         return current_pid != pid ? -ESRCH : 0;
 }
+
+int pidfd_get_inode_id(int fd, uint64_t *ret) {
+        static int cached_supported = -1;
+
+        assert(fd >= 0);
+
+        if (cached_supported < 0) {
+                cached_supported = fd_is_fs_type(fd, PID_FS_MAGIC);
```

Comment:
Hm? Why cached?? The result depends on the input fd, no?

Please add simple unit tests for the function.

Reply:
Then, that's not a good idea to merge this with `pidfd_inode_ids_supported()`.
When this is exposed, an arbitrary fd may be passed mistakenly, and we cannot set any assertion for the misuse.
So, keep `pidfd_inode_ids_supported()` as is, and call it in this function to check if the functionality is supported. The result is cached. Hence should not impact performance.

Reply:
>  When this is exposed, an arbitrary fd may be passed mistakenly, and we cannot set any assertion for the misuse.

But what this function does is essentially getting inode number. So we really cannot prevent such case (unless `fd_is_fs_type()` is checked for each and every invocation. `ioctl`s, similarly, can mean completely different stuff on different fds. We generally do not protect against totally non-sensical mistakes like this?

Reply:
Even though, it is not necessary to introduce any further pitfall.
The problem here is, if this is called first time with a wrong fd, then the result is cached, and the inode number cannot be used by any later calls. And, it is hard to notice the misuse here. We can avoid such misuse here, then why not to do?

Moreover, I'd like to check if the passed fd a pidfd before calling ioctl()s in other places. That can be done if `PID_FS_MAGIC` is supported.

Reply:
All these seem too much churn than it's worth... We have limited call sites, just like all other uses of ioctl. And this isn't particularly cheap. Just like we assert on `fd >= 0`, not `fd_validate()`.